src/Microsoft.Health.Fhir.Core/Features/Operations/Import/ImportOrchestratorTask.cs

```diff
@@ -0,0 +1,505 @@
+                    }
+                }
+                else if (taskResultData.Result == TaskResult.Fail)
+                {
+                    throw new ImportProcessingException(string.Format("Failed to process file: {0}. {1}", completedResourceUri, taskResultData));
+                }
+                else if (taskResultData.Result == TaskResult.Canceled)
+                {
+                    throw new OperationCanceledException(taskResultData.ResultData);
+                }
+            }
+        }
+
+        private async Task<List<Uri>> MonitorRunningTasksAsync(List<(Uri resourceUri, TaskInfo taskInfo)> runningTasks, CancellationToken cancellationToken)
+        {
+            List<Uri> completedTaskResourceUris = new List<Uri>();
+
+            foreach ((Uri runningResourceUri, TaskInfo runningTaskInfo) in runningTasks)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException();
+                }
+
+                TaskInfo latestTaskInfo = await _taskManager.GetTaskAsync(runningTaskInfo.TaskId, cancellationToken);
+
+                _orchestratorTaskContext.DataProcessingTasks[runningResourceUri] = latestTaskInfo;
+                if (latestTaskInfo.Status == TaskStatus.Completed)
+                {
+                    completedTaskResourceUris.Add(runningResourceUri);
+                }
+            }
+
+            return completedTaskResourceUris;
+        }
+
+        private async Task CancelProcessingTasksAsync()
+        {
+            List<string> runningTaskIds = new List<string>();
+
+            if ((_orchestratorTaskContext?.DataProcessingTasks?.Count ?? 0) == 0)
+            {
+                // No data processing task created.
+                return;
+            }
+
+            foreach (TaskInfo taskInfo in _orchestratorTaskContext.DataProcessingTasks.Values)
+            {
+                try
+                {
+                    TaskInfo taskInfoFromServer = await _taskManager.GetTaskAsync(taskInfo.TaskId, CancellationToken.None);
+
+                    if (taskInfoFromServer != null)
+                    {
+                        await _taskManager.CancelTaskAsync(taskInfo.TaskId, CancellationToken.None);
+                        runningTaskIds.Add(taskInfo.TaskId);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogWarning(ex, "failed to cancel task {0}", taskInfo.TaskId);
+                }
+            }
+
+            // Wait task cancel for WaitRunningTaskCancelTimeoutInSec
+            await WaitRunningTaskCompleteAsync(runningTaskIds);
+        }
+
+        private async Task WaitRunningTaskCompleteAsync(List<string> runningTaskIds)
```

Comment:
This method looks like it could get into an infinite loop if it can't get info for a task.

Reply:
the task should be created first then added to running task queue, so it should not happen. Make it clear looks like also remove task id could be more readable.